pkg/controller/control.go

```diff
@@ -539,27 +542,82 @@ func (c *Controller) Start(addresses ...string) error {
 	// For more details, see the following url:
 	// https://github.com/longhorn/longhorn/issues/1235
 	if !c.isUpgrade {
-		revisionCounters := make(map[string]int64)
-		for _, r := range c.replicas {
-			// The related backend is nil if the mode is ERR
-			if r.Mode == types.ERR {
-				continue
+		// Find best replica for salvge recovering, based on lastModifyTime
+		// and HeadFileSize.
+		if c.salvageRequested {
+			replicaCandidates := make(map[types.Replica]*types.ReplicaInfo)
+			var lastModifyTime int64
+			for _, r := range c.replicas {
+				repClient, err := replicaClient.NewReplicaClient(r.Address)
+				if err != nil {
+					return err
+				}
+
+				rep, err := repClient.GetReplica()
+				if err != nil {
+					return err
+				}
+				replicaCandidates[r] = rep
+				if lastModifyTime == 0 ||
+					rep.LastModifyTime < lastModifyTime {
+					lastModifyTime = rep.LastModifyTime
+				}
 			}
-			counter, err := c.backend.GetRevisionCounter(r.Address)
-			if err != nil {
-				return err
+
+			if len(replicaCandidates) != 0 {
+				var bestCandidate types.Replica
+				lastTime := time.Unix(0, lastModifyTime)
+				var largestSize int64
+				for r, replicaInfo := range replicaCandidates {
+					t := time.Unix(0, replicaInfo.LastModifyTime)
+					// Any replica within 5 seconds before lastModifyTime
+					// can be good replica.
+					if t.Add(time.Second * 5).After(lastTime) {
+						if replicaInfo.HeadFileSize >= largestSize {
+							bestCandidate = r
+						}
+					}
+				}
+
+				if bestCandidate == (types.Replica{}) {
+					logrus.Errorf("BUG: Should find one candidate for salvage")
```

Comment:
If there is no replica for the salvage, we may need to return an error here, right? Then longhorn manager shouldn't retry the auto salvage.

Similar below.

Reply:
Sure, will fix them.